File: protocol-generator/java/src/main/java/org/entitycore/protocol/peer/Outcome.java
package org.entitycore.protocol.peer;

import java.util.List;

/**
 * A handler outcome: a status, a result entity, and any protocol entities to carry in
 * the response envelope's {@code included} (§3.1) — caps, peer identities, signatures.
 */
public record Outcome(int status, Entity result, List<Envelope.Included> included) {

    public static Outcome ok(Entity result) {
        return new Outcome(200, result, List.of());
    }

    public static Outcome ok(Entity result, List<Envelope.Included> included) {
        return new Outcome(200, result, included);
    }

    public static Outcome err(int status, String code) {
        return new Outcome(status, Wire.errorResult(code, null), List.of());
    }

    public static Outcome err(int status, String code, String message) {
        return new Outcome(status, Wire.errorResult(code, message), List.of());
    }
}

File: protocol-generator/php/src/Outcome.php
<?php

declare(strict_types=1);

namespace EntityCore;

/**
 * A handler result: the §3.3 status + the result entity + the §3.1 `included`
 * authority the result carries (the §4.4 grant mint travels here). The
 * recoverable protocol path is this sealed-result seam (NOT a throw — exceptions
 * are reserved for the unrecoverable transport/codec boundary; profile
 * [error_model]).
 */
final class Outcome
{
    /**
     * @param list<array{hash:string,entity:Entity}> $included
     */
    private function __construct(
        public readonly int $status,
        public readonly Entity $result,
        public readonly array $included,
    ) {
    }

    /**
     * A 200 success.
     *
     * @param list<array{hash:string,entity:Entity}> $included
     */
    public static function ok(Entity $result, array $included = []): self
    {
        return new self(200, $result, $included);
    }

    /** An error status with a system/protocol/error result body. */
    public static function err(int $status, string $code, ?string $message = null): self
    {
        return new self($status, Wire::errorResult($code, $message), []);
    }
}
